hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

```diff
@@ -1817,10 +1839,10 @@ protected LocatedBlock nextBlockOutputStream() throws IOException {
       nodes = lb.getLocations();
       nextStorageTypes = lb.getStorageTypes();
       nextStorageIDs = lb.getStorageIDs();
-
+      setPipeline(lb);
       // Connect to first DataNode in the list.
       success = createBlockOutputStream(nodes, nextStorageTypes, nextStorageIDs,
-          0L, false);
+          0L, false) || setupPipelineForAppendOrRecovery();
```

Comment:
By this, if your first attempt to create a stream with 3 nodes fails, you immediately fallback to dropping the Datanode.

Shouldn't we try to get a node replacement like before & if those attempts fail then comeback on dropping logic? So, if there is just one node problem (not a rack level), before this change the stream would have got a replacement but now it won't?

If I catch it right, it is a behaviour change, right?

Reply:
> If I catch it right, it is a behaviour change, right?

I agree this is a behavior change but the current behavior is buggy. `dtpReplaceDatanodeOnFailureReplication`  configuration is NOT honored during the `PIPELINE_SETUP_CREATE` phase. Thats why this jira was created. 
After this patch, we will handle datanode failure during  `PIPELINE_SETUP_CREATE` in the same way as `DATA_STREAMING` phase and the application (in this case hbase) will experience consistent behavior using hdfs client.
Cc @ritegarg  @ayushtkn

Reply:
hmm, to me it looks fine only, I am not sure about any legacy reasons or any use case which can break due to this. If other folks are ok, should be fine.

@Hexiaoqiao / @tasanuma mind giving a check

Reply:
I haven't looked into the PR in detail, but it makes sense to me that PIPELINE_SETUP_CREATE should also consider the `dtpReplaceDatanodeOnFailureReplication`. If I understand correctly, this change won't affect users who have set `dfs.client.block.write.replace-datanode-on-failure.min-replication(=dtpReplaceDatanodeOnFailureReplication)=0`, which is the default setting, so I think it's fairly safe.